### PR TITLE
Use fst for symbol search

### PR DESCRIPTION
**Pull request by matklad**
WIP to get the feedback on the general idea.

This switches from prefix_trie to fst by @burntsushi, which allows for much more powerful API:

* search symbols either by prefix or by subsequence. This should fix the problem that `ctrl+T > vfs` in VS Code does not find `range_from_vfs_file`.

* limit the number of results returned. This should fix the problem of VS Code freezing after pressing `ctrl+T` and not typing a query string (which currently retrieves all of the symbols)

* support for streaming API. Now it is possible to get a huge list of symbols in small chunks, so that symbol search in the editor can give results instantly, and then load more results as user scrolls the list. Note that LSP/[VS Code API](https://code.visualstudio.com/docs/extensionAPI/vscode-api#WorkspaceSymbolProvider)  in this area is suboptimal: symbol search returns a `Future<Vec<Symbol>>` and not a `Stream<Symbol>`. Hopefully that will be fixed one day.

### CODE REVIEW

**Comment by matklad** [comment]
BTW, VS Code docs explicitly advice against using prefix-match :) 

>A good rule of thumb is to match case-insensitive and to simply check that the characters of query appear in their order in a candidate symbol. Don't use prefix, substring, or similar strict matching.

https://code.visualstudio.com/docs/extensionAPI/vscode-api#WorkspaceSymbolProvider

File: src/analysis.rs [hunk 1]
```diff
@@ -46,7 +46,7 @@ pub struct PerCrateAnalysis {
     pub defs_per_file: HashMap<PathBuf, Vec<Id>>,
     pub children: HashMap<Id, HashSet<Id>>,
     pub def_names: HashMap<String, Vec<Id>>,
-    pub def_trie: Trie<String, Vec<Id>>,
+    // pub def_trie: Trie<String, Vec<Id>>,
```
Comment (nrc) [inline src/analysis.rs]: Could you remove this rather than comment it out please?

File: src/fstq.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
+use fst;
```
Comment (nrc) [inline src/fstq.rs]: Could you add some comments explaining what this module is for?

File: src/analysis.rs [hunk 3]
```diff
@@ -147,6 +147,17 @@ impl PerCrateAnalysis {
...
 }
 
+pub struct MatchingDefsQuery {
+    pub kind: MatchingDefsQueryKind,
+    pub limit: usize,
+    pub greater_than: String,
+}
+
+pub enum MatchingDefsQueryKind {
+    Preifx(String),
```
Comment (nrc) [inline src/analysis.rs]: typo: Prefix

File: src/analysis.rs [hunk 4]
```diff
@@ -284,21 +295,44 @@ impl Analysis {
...
-                }).collect();
-            Some(defs)
+        self.matching_defs_q(MatchingDefsQuery {
+            kind: MatchingDefsQueryKind::Preifx(stem.to_lowercase()),
+            limit: <usize>::max_value(),
+            greater_than: String::new(),
         })
     }
 
+    pub fn matching_defs_q(&self, query: MatchingDefsQuery) -> Vec<Def> {
```
Comment (nrc) [inline src/analysis.rs]: s/_q/_query please

Comment (nrc) [reply src/analysis.rs]: or _with_query

**Comment by matklad** [comment]
Cleaned-up the implementation, should be ready for the review now

**Comment by matklad** [comment]
Might be a good idea to cut a release after this PR.

**Comment by nrc** [comment]
Thanks!

**Comment by nrc** [comment]
Publishd 0.14.0

**Pull request merged**